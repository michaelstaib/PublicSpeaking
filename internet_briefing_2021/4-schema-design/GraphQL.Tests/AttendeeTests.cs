using System;
using System.Threading.Tasks;
using ConferencePlanner.GraphQL;
using ConferencePlanner.GraphQL.Attendees;
using ConferencePlanner.GraphQL.Data;
using ConferencePlanner.GraphQL.Sessions;
using ConferencePlanner.GraphQL.Speakers;
using ConferencePlanner.GraphQL.Tracks;
using ConferencePlanner.GraphQL.Types;
using HotChocolate;
using HotChocolate.Execution;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using Xunit;

namespace GraphQL.Tests
{
    public class AttendeeTests
    {
        [Fact]
        public void Attendee_Schema_Changed()
        {
            var services = new ServiceCollection();
            services.AddDbContextPool<ApplicationDbContext>(
                options => options.UseInMemoryDatabase("Data Source=conferences.db"));

             SchemaBuilder.New()
                .AddServices(services.BuildServiceProvider())
                .AddQueryType(d => d.Name("Query"))
                    .AddType<AttendeeQueries>()
                .AddMutationType(d => d.Name("Mutation"))
                    .AddType<AttendeeMutations>()
                .AddType<AttendeeType>()
                .AddType<SessionType>()
                .AddType<SpeakerType>()
                .AddType<TrackType>()
                .EnableRelaySupport()
                .Create()
                .ToString()
                .MatchSnapshot();
        }
        
        [Fact]
        public async Task RegisterAttendee()
        {
            // arrange
            IServiceProvider services = new ServiceCollection()
                .AddDbContextPool<ApplicationDbContext>(
                    options => options.UseInMemoryDatabase("Data Source=conferences.db"))
                .AddGraphQL()
                    .AddQueryType(d => d.Name("Query"))
                        .AddType<AttendeeQueries>()
                    .AddMutationType(d => d.Name("Mutation"))
                        .AddType<AttendeeMutations>()
                    .AddType<AttendeeType>()
                    .AddType<SessionType>()
                    .AddType<SpeakerType>()
                    .AddType<TrackType>()
                    // .EnableRelaySupport()
                .Services
                .BuildServiceProvider();
            
            // act
            IExecutionResult result = await services.ExecuteRequestAsync(
                QueryRequestBuilder.New()
                    .SetQuery(@"
                        mutation RegisterAttendee {
                            registerAttendee(
                                input: {
                                    emailAddress: ""michael@chillicream.com""
                                        firstName: ""michael""
                                        lastName: ""staib""
                                        userName: ""michael3""
                                    }) 
                            {
                                attendee {
                                    id
                                }
                            }
                        }")
                    .Create());
            
            // assert
            result.MatchSnapshot();
        }
    }
}
