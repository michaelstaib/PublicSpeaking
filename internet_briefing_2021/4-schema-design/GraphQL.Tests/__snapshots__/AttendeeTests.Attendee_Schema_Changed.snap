schema {
  query: Query
  mutation: Mutation
}

"The node interface is implemented by entities that have a global unique identifier."
interface Node {
  id: ID!
}

type Attendee implements Node {
  emailAddress: String
  firstName: String!
  id: ID!
  lastName: String!
  sessions: [Session!]!
  userName: String!
}

"A connection to a list of items."
type AttendeeConnection {
  "A list of edges."
  edges: [AttendeeEdge!]
  "A flattened list of the nodes."
  nodes: [Attendee]
  "Information to aid in pagination."
  pageInfo: PageInfo!
  totalCount: Int!
}

"An edge in a connection."
type AttendeeEdge {
  "A cursor for use in pagination."
  cursor: String!
  "The item at the end of the edge."
  node: Attendee
}

type CheckInAttendeePayload {
  attendee: Attendee
  clientMutationId: String
  errors: [UserError!]!
  session: Session
}

type Mutation {
  checkInAttendee(input: CheckInAttendeeInput!): CheckInAttendeePayload!
  registerAttendee(input: RegisterAttendeeInput!): RegisterAttendeePayload!
}

"Information about pagination in a connection."
type PageInfo {
  "When paginating forwards, the cursor to continue."
  endCursor: String
  "Indicates whether more edges exist following the set defined by the clients arguments."
  hasNextPage: Boolean!
  "Indicates whether more edges exist prior the set defined by the clients arguments."
  hasPreviousPage: Boolean!
  "When paginating backwards, the cursor to continue."
  startCursor: String
}

type Query {
  attendeeById(id: ID!): Attendee!
  attendees(after: String before: String first: PaginationAmount last: PaginationAmount): AttendeeConnection
  attendeesById(ids: [ID]): [Attendee!]!
  node(id: ID!): Node
}

type RegisterAttendeePayload {
  attendee: Attendee
  clientMutationId: String
  errors: [UserError!]!
}

type Session implements Node {
  abstract: String
  attendees: [Attendee!]!
  duration: TimeSpan!
  endTime: DateTime
  id: ID!
  speakers: [Speaker!]!
  startTime: DateTime
  title: String!
  track: Track
  trackId: ID
}

"A connection to a list of items."
type SessionConnection {
  "A list of edges."
  edges: [SessionEdge!]
  "A flattened list of the nodes."
  nodes: [Session!]
  "Information to aid in pagination."
  pageInfo: PageInfo!
  totalCount: Int!
}

"An edge in a connection."
type SessionEdge {
  "A cursor for use in pagination."
  cursor: String!
  "The item at the end of the edge."
  node: Session!
}

type Speaker implements Node {
  bio: String
  id: ID!
  name: String!
  sessions: [Session!]!
  webSite: String
}

type Track implements Node {
  id: ID!
  name: String!
  sessions(after: String before: String first: PaginationAmount last: PaginationAmount): SessionConnection
}

type UserError {
  code: String!
  message: String!
}

input CheckInAttendeeInput {
  attendeeId: ID!
  clientMutationId: String
  sessionId: ID!
}

input RegisterAttendeeInput {
  clientMutationId: String
  emailAddress: String!
  firstName: String!
  lastName: String!
  userName: String!
}

"The `Boolean` scalar type represents `true` or `false`."
scalar Boolean

"The `DateTime` scalar represents an ISO-8601 compliant date time type."
scalar DateTime

"The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `\"4\"`) or integer (such as `4`) input value will be accepted as an ID."
scalar ID

"The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1."
scalar Int

scalar PaginationAmount

"The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text."
scalar String

scalar TimeSpan
